File: src/mithrillog/ingestion/dedupe.py
from __future__ import annotations

import hashlib
import math
import os
from dataclasses import dataclass, field
from typing import Iterable


def _hashes(value: bytes, num_hashes: int) -> Iterable[int]:
    digest = hashlib.sha256(value).digest()
    for i in range(num_hashes):
        start = i * 4
        chunk = digest[start : start + 4]
        if len(chunk) < 4:
            chunk = (chunk + hashlib.sha1(chunk).digest())[:4]
        yield int.from_bytes(chunk, "little", signed=False)


@dataclass
class BloomDeduper:
    """
    Tiny bloom filter for minute buckets.
    """

    capacity: int = 1_000_000
    error_rate: float = 1e-4
    seed: int = field(default_factory=lambda: int.from_bytes(os.urandom(4), "little"))

    def __post_init__(self) -> None:
        m = -self.capacity * math.log(self.error_rate) / (math.log(2) ** 2)
        self.num_bits = max(8, int(m))
        self.num_hashes = max(2, int(self.num_bits / self.capacity * math.log(2)))
        self.bitset = bytearray(math.ceil(self.num_bits / 8))

    def _positions(self, payload: bytes) -> list[int]:
        salted = self.seed.to_bytes(4, "little") + payload
        return [pos % self.num_bits for pos in _hashes(salted, self.num_hashes)]

    def seen(self, payload: bytes) -> bool:
        positions = self._positions(payload)
        present = all(self.bitset[p // 8] & (1 << (p % 8)) for p in positions)
        if not present:
            for p in positions:
                self.bitset[p // 8] |= 1 << (p % 8)
        return present

    def reset(self) -> None:
        for idx in range(len(self.bitset)):
            self.bitset[idx] = 0


@dataclass
class ReservoirSampler:
    """
    Per-key reservoir sampler keeping representative examples.
    """

    size: int = 200

    def __post_init__(self) -> None:
        self._reservoir: dict[str, list[dict]] = {}
        self._counts: dict[str, int] = {}
        self._pattern_counts: dict[str, int] = {}

    def add(self, key: str, pattern: str, event: dict) -> None:
        count = self._counts.get(key, 0) + 1
        self._counts[key] = count
        self._pattern_counts[pattern] = self._pattern_counts.get(pattern, 0) + 1
        reservoir = self._reservoir.setdefault(key, [])
        if len(reservoir) < self.size:
            reservoir.append(event)
            return
        import random

        index = random.randint(0, count - 1)
        if index < self.size:
            reservoir[index] = event

    def register_duplicate(self, pattern: str) -> None:
        self._pattern_counts[pattern] = self._pattern_counts.get(pattern, 0) + 1

    def snapshot(self) -> dict[str, list[dict]]:
        return {key: values.copy() for key, values in self._reservoir.items()}

    def totals(self) -> dict[str, int]:
        return self._pattern_counts.copy()

    def reset(self) -> None:
        self._reservoir.clear()
        self._counts.clear()
        self._pattern_counts.clear()

File: src/ingester-rs/src/dedupe.rs
use bloomfilter::Bloom;
use rand::Rng;
use std::collections::HashMap;
use serde_json::Value;

pub struct BloomDeduper {
    filter: Bloom<Vec<u8>>,
}

impl BloomDeduper {
    pub fn new(capacity: usize, error_rate: f64) -> Self {
        // Bloom::new_for_fp_rate(items_count, fp_rate)
        Self {
            filter: Bloom::new_for_fp_rate(capacity, error_rate),
        }
    }

    pub fn seen(&mut self, key: &[u8]) -> bool {
        // check_and_set returns true if item was already present
        self.filter.check_and_set(&key.to_vec())
    }

    pub fn reset(&mut self) {
        self.filter.clear();
    }
}

pub struct ReservoirSampler {
    size: usize,
    reservoir: HashMap<String, Vec<Value>>,
    counts: HashMap<String, usize>,
    pattern_counts: HashMap<String, usize>,
}

impl ReservoirSampler {
    pub fn new(size: usize) -> Self {
        Self {
            size,
            reservoir: HashMap::new(),
            counts: HashMap::new(),
            pattern_counts: HashMap::new(),
        }
    }

    pub fn add(&mut self, key: String, pattern: String, event: Value) {
        let count = self.counts.entry(key.clone()).or_insert(0);
        *count += 1;

        let p_count = self.pattern_counts.entry(pattern.clone()).or_insert(0);
        *p_count += 1;

        let samples = self.reservoir.entry(key).or_insert_with(Vec::new);
        
        if samples.len() < self.size {
            samples.push(event);
        } else {
            let mut rng = rand::thread_rng();
            let index = rng.gen_range(0..*count);
            if index < self.size {
                samples[index] = event;
            }
        }
    }

    pub fn register_duplicate(&mut self, pattern: String) {
        let count = self.pattern_counts.entry(pattern).or_insert(0);
        *count += 1;
    }

    pub fn totals(&self) -> HashMap<String, usize> {
        self.pattern_counts.clone()
    }

    pub fn reset(&mut self) {
        self.reservoir.clear();
        self.counts.clear();
        self.pattern_counts.clear();
    }

    pub fn snapshot(&self) -> HashMap<String, Vec<Value>> {
        self.reservoir.clone()
    }
}
